Don't require `deps.edn` for local scripts

teodorlu opened the pull request:
Please answer the following questions and leave the below in as part of your PR.

- [x] This PR corresponds to an [issue with a clear problem statement](https://github.com/babashka/babashka/blob/master/doc/dev.md#start-with-an-issue-before-writing-code).
  - https://github.com/babashka/bbin/issues/78

teodorlu committed:
Don't require `deps.edn` for local scripts

teodorlu commented:
With the bbin changes in 64025cc, I can install the `fail1` script from https://github.com/teodorlu/bbin-testdata/tree/e8c5a94f042a73076010d04948139c2590a313ab/fail1/ without issues:

```
$ git rev-parse HEAD
e8c5a94f042a73076010d04948139c2590a313ab
$ pwd
/Users/teodorlu/dev/teodorlu/bbin-testdata/fail1
$ bbin-dev install .

Starting install...

bin    location                                               
─────  ───────────────────────────────────────────────────────
fail1  file:///Users/teodorlu/dev/teodorlu/bbin-testdata/fail1

Install complete.

$ fail1
{:script "fail1", :args nil}
```

borkdude commented:
@teodorlu Cool. Perhaps bb.edn should be prioritized over deps.edn, what do you think?

rads commented:
Thanks for the PR, taking a closer look now.

teodorlu commented:
> @teodorlu Cool. Perhaps bb.edn should be prioritized over deps.edn, what do you think?

Yeah, that's probably the right choice! I'll make the change.

teodorlu commented:
> Thanks for the PR, taking a closer look now.

Note: currently, this is just a start! But I wanted to put some code up so that we can discuss real code rather than hypothetical code.

- I only changed the behavior for installing script from a local directory
- I haven't added any tests yet

teodorlu committed:
Prioritize `bb.edn` over `deps.edn`

Rationale: `bb.edn` is _meant_ to control babashka-things, and may
_optionally refer_ to `deps.edn` for additional dependencies.

teodorlu commented:
> Makes sense. The main thing I'm thinking about is backwards-compatibility. Whatever change we make, I want to ensure we're only adding behavior and existing users don't have to do anything differently.

100 % agreed! Let's not break userspace.

teodorlu committed:
Add test for local-dir-template-str

teodorlu commented:
Some observations while working on this code, probably out of scope for this PR:

1. I wonder if creating the `tmp-edn` file is actually required. It seems it's not required when a `deps.edn` file or a `bb.edn` file exists.

    ```clojure
    (def tmp-edn
      (doto (fs/file (fs/temp-dir) (str (gensym \"bbin\")))
        (spit (str \"{:deps {\" script-lib script-coords \"}}\"))
        (fs/delete-on-exit)))
    ```

2. I wonder if script shims would be easier to work with as Clojure data rather than string template literals, getting the Clojure code right inside the string literals can be a bit hard.

borkdude committed:
Merge branch 'main' into do-not-require-deps-edn-file-for-installing-scripts

rads commented:
@teodorlu: I agree that those are out-of-scope, but you're welcome to create issues for those and anything else that comes up.

rads commented:
I'm planning to fix this issue in a separate PR so I'm going to close this one.
